tests/components/modbus/conftest.py

```diff
@@ -37,61 +40,84 @@ def __init__(self, register_words):
         self.bits = register_words
 
 
-async def setup_base_test(
+async def base_test(
     sensor_name,
     hass,
-    use_mock_hub,
     data_array,
     entity_domain,
     scan_interval,
+    register_words,
+    expected,
+    method_discovery=False,
 ):
-    """Run setup device for given config."""
+    """Run test on device for given config."""
 
     # Full sensor configuration
-    config = {
-        entity_domain: {
-            CONF_PLATFORM: "modbus",
-            CONF_SCAN_INTERVAL: scan_interval,
-            **data_array,
+    if method_discovery:
+        config = {
+            DOMAIN: {
+                CONF_NAME: DEFAULT_HUB,
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTest",
+                CONF_PORT: 5001,
+                **data_array,
+            },
+        }
+    else:
+        config = {
+            DOMAIN: {
+                CONF_NAME: DEFAULT_HUB,
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTest",
+                CONF_PORT: 5001,
+            },
+        }
+        configDeviceOLD = {
+            entity_domain: {
+                CONF_PLATFORM: DOMAIN,
+                CONF_SCAN_INTERVAL: scan_interval,
+                **data_array,
+            }
         }
-    }
 
-    # Initialize sensor
+    # mock timer and add modbus platform with devices (new config)
+    # first add modbus platform, then devices (old config)
     now = dt_util.utcnow()
     with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
-        assert await async_setup_component(hass, entity_domain, config)
+        # setup modbus platform
+        if method_discovery:
+            await async_load_platform(
+                hass, entity_domain, DOMAIN, config[DOMAIN], config
+            )
+        else:
+            # setup modbus old style
+            assert await async_setup_component(hass, DOMAIN, config)
+            await hass.async_block_till_done()
+
+            # setup component old style
+            assert await async_setup_component(
+                hass,
+                entity_domain,
+                configDeviceOLD,
+            )
+
         await hass.async_block_till_done()
 
+    # Setup inputs for the sensor
+    read_result = ReadResult(register_words)
+    hub: ModbusHub = hass.data[DOMAIN][DEFAULT_HUB]
```

Comment:
We shouldn't interact with details in the tests. Patch the library as needed to avoid I/O and use the core interfaces to start the tests.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Reply:
I see your point, however if I use Hass.states to set the state, the whole idea of testing if void, because what we want to test is that the result from the pymodbus library is correctly converted to a state.

hub: ModbusHub = hass.data[DOMAIN][DEFAULT_HUB] retrieves a MagicMock and read_result is used as return value in a function in the MagicMock, so MagicMock replaces the pymodbus library (actually the class we added on top of the library).

Due to MagicMock there are no IO not is pymodbus ever called in test.

If I use Hass.states to set state, there are no way to test the conversion from "raw" input to a well defined state.
(remark this is not new code, this have been in here since the beginning).

Reply:
We don't want to set fake states with hass.states.

Set up the integration but don't interact with any integration details, like accessing hass.data. `hass.data` holds internal objects that we shouldn't rely on in our tests.